internal/federationdomain/resolvedprovider/resolvedgithub/resolved_github_provider.go

```diff
@@ -91,18 +94,52 @@ func (p *FederationDomainResolvedGitHubIdentityProvider) Login(
 }
 
 func (p *FederationDomainResolvedGitHubIdentityProvider) LoginFromCallback(
-	_ context.Context,
-	_ string,
-	_ pkce.Code,
-	_ nonce.Nonce,
-	_ string,
+	ctx context.Context,
+	authCode string,
+	_ pkce.Code, // GitHub does not support PKCE, see https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps
+	_ nonce.Nonce, // GitHub does not support OIDC, therefore there is no ID token that could contain the "nonce".
+	redirectURI string,
 ) (*resolvedprovider.Identity, *resolvedprovider.IdentityLoginExtras, error) {
-	return nil, nil, errors.New("function LoginFromCallback not yet implemented for GitHub IDP")
+	accessToken, err := p.Provider.ExchangeAuthcode(ctx, authCode, redirectURI)
+	if err != nil {
+		plog.WarningErr("error exchanging GitHub authcode", err, "upstreamName", p.Provider.GetName())
+		return nil, nil, httperr.Wrap(http.StatusBadGateway,
+			fmt.Sprintf("failed to exchange authcode using GitHub API: %s", err.Error()),
+			err,
+		)
+	}
+
+	user, err := p.Provider.GetUser(ctx, accessToken, p.GetDisplayName())
+	if err != nil {
+		return nil, nil, httperr.Wrap(http.StatusUnprocessableEntity,
+			fmt.Sprintf("failed to get user info from GitHub API: %s", err.Error()),
+			err,
+		)
+	}
+
+	return &resolvedprovider.Identity{
+			UpstreamUsername:  user.Username,
+			UpstreamGroups:    user.Groups,
+			DownstreamSubject: user.DownstreamSubject,
+			IDPSpecificSessionData: &psession.GitHubSessionData{
+				UpstreamAccessToken: accessToken,
+			},
+		},
+		&resolvedprovider.IdentityLoginExtras{
+			DownstreamAdditionalClaims: nil, // not using this for GitHub
+			Warnings:                   nil, // not using this for GitHub
+		},
+		nil // no error
 }
 
 func (p *FederationDomainResolvedGitHubIdentityProvider) UpstreamRefresh(
 	_ context.Context,
-	_ *resolvedprovider.Identity,
-) (refreshedIdentity *resolvedprovider.RefreshedIdentity, err error) {
-	return nil, errors.New("function UpstreamRefresh not yet implemented for GitHub IDP")
+	identity *resolvedprovider.Identity,
+) (*resolvedprovider.RefreshedIdentity, error) {
+	// TODO: actually implement refresh. this is just a placeholder that will make refresh always succeed.
```

Comment:
This TODO is for the next user story. We do not need to do this before merge for this user story.